### PR TITLE
feat(1994): [2] Filtering for ~release and ~tag triggers

### PR DESCRIPTION
**Pull request by ibu1224**
## Context
feat https://github.com/screwdriver-cd/screwdriver/issues/1994

## Objective
This PR will allow you to filter by release triggers and tag triggers.
スクリーンショット 2020-06-15 11 49 29
Filter by 'releaseName' and 'ref' in the webhook in case of tag trigger and release trigger.

## References

https://github.com/screwdriver-cd/screwdriver/issues/1994
[1]https://github.com/screwdriver-cd/models/pull/460
[3] https://github.com/screwdriver-cd/guide/pull/386
## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/webhooks/index.js [hunk 1]
```diff
@@ -22,17 +64,17 @@ const DEFAULT_MAX_BYTES = 1048576;
  * @param {String} pipelineBranch  The pipeline branch
```
Comment (tkyi) [inline plugins/webhooks/index.js:64]: need to add new param jsdoc

Comment (ibu1224) [reply plugins/webhooks/index.js]: Thank you for review. 
I add param in jsdoc.

**Comment by ibu1224** [comment]
I was just discussing it with the members of YJ.
I implemented the following `~commit` cases and differently.
スクリーンショット 2020-06-12 15 09 56

My implementation of the `~release` case
スクリーンショット 2020-06-12 15 12 50

I would modify it to work the same way as `~commit`.

File: plugins/webhooks/index.js [hunk 2]
```diff
@@ -14,25 +14,49 @@ const CHECKOUT_URL_SCHEMA = schema.config.regex.CHECKOUT_URL;
...
+function determineStartFrom(action, type, targetBranch, pipelineBranch, releaseName, tagName, isReleaseOrTagFiltering) {
     let startFrom;
 
     if (type && type === 'pr') {
         startFrom = '~pr';
     } else {
         switch (action) {
             case 'release':
-                return '~release';
+                return isReleaseOrTagFiltering && action === 'release' ? `~release:${releaseName}` : '~release';
```
Comment (wahapo) [inline plugins/webhooks/index.js:57]: I think it is better to check if `releaseName` and `tagName` are undefined.

File: plugins/webhooks/index.js [hunk 3]
```diff
@@ -264,7 +299,7 @@ async function triggeredPipelines(pipelineFactory, scmConfig, branch, type, acti
     );
 
     pipelinesOnOtherBranch = pipelinesOnOtherBranch.filter(p =>
-        hasTriggeredJob(p, determineStartFrom(action, type, branch, null))
+        hasTriggeredJob(p, determineStartFrom(action, type, branch, null, releaseName, tagName))
```
Comment (wahapo) [inline plugins/webhooks/index.js:302]: Is the argument of `isReleaseOrTagFiltering` unnecessary?

File: plugins/webhooks/index.js (hunk 2 above)
Comment (wahapo) [inline plugins/webhooks/index.js:57]: In this case you can omit `action === 'release'`.

Comment (ibu1224) [reply plugins/webhooks/index.js:57]: Add undefined check.

Comment (ibu1224) [reply plugins/webhooks/index.js:57]: You are correct.
I fix it.

File: plugins/webhooks/index.js (hunk 3 above)
Comment (ibu1224) [reply plugins/webhooks/index.js:302]: I add `isReleaseOrTagFiltering`.

File: plugins/webhooks/index.js [hunk 4]
```diff
@@ -14,25 +14,61 @@ const CHECKOUT_URL_SCHEMA = schema.config.regex.CHECKOUT_URL;
...
 const DEFAULT_MAX_BYTES = 1048576;
 
+/**
+ * Check if the tag or release filtering or not
+ * @param {String}    action          SCM webhook action type
+ * @param {Array}     workflowGraph   pipeline workflowGraph
+ * @returns {Boolean} isFilterlingEnabled
+ */
+function isReleaseOrTagFilteringEnabled(action, workflowGraph) {
+    let isFilterlingEnabled = true;
```
Comment (tkyi) [inline plugins/webhooks/index.js:24]:
```suggestion
    let isFilteringEnabled = true;
```

File: plugins/webhooks/index.js [hunk 5]
```diff
@@ -14,25 +14,61 @@ const CHECKOUT_URL_SCHEMA = schema.config.regex.CHECKOUT_URL;
 const CHECKOUT_URL_SCHEMA_REGEXP = new RegExp(CHECKOUT_URL_SCHEMA);
 const DEFAULT_MAX_BYTES = 1048576;
 
+/**
+ * Check if the tag or release filtering or not
+ * @param {String}    action          SCM webhook action type
+ * @param {Array}     workflowGraph   pipeline workflowGraph
+ * @returns {Boolean} isFilterlingEnabled
```
Comment (tkyi) [inline plugins/webhooks/index.js:21]:
```suggestion
 * Check if tag or release filtering is enabled or not
 * @param {String}    action          SCM webhook action type
 * @param {Array}     workflowGraph   pipeline workflowGraph
 * @returns {Boolean} isFilteringEnabled
```

File: plugins/webhooks/index.js [hunk 6]
```diff
@@ -14,25 +14,61 @@ const CHECKOUT_URL_SCHEMA = schema.config.regex.CHECKOUT_URL;
...
+ */
+function isReleaseOrTagFilteringEnabled(action, workflowGraph) {
+    let isFilterlingEnabled = true;
+
+    workflowGraph.edges.forEach(edge => {
+        const releaseOrTagRegExp = action === 'release' ? new RegExp('^~(release)$') : new RegExp('^~(tag)$');
+
+        if (releaseOrTagRegExp) {
+            if (edge.src.match(releaseOrTagRegExp)) {
+                isFilterlingEnabled = false;
```
Comment (tkyi) [inline plugins/webhooks/index.js:31]:
```suggestion
                isFilteringEnabled = false;
```

File: plugins/webhooks/index.js [hunk 7]
```diff
@@ -14,25 +14,61 @@ const CHECKOUT_URL_SCHEMA = schema.config.regex.CHECKOUT_URL;
...
- * @param {String} targetBranch    The branch against which commit is pushed
- * @param {String} pipelineBranch  The pipeline branch
- * @returns {String}               startFrom
+ * @param {String}   action                    SCM webhook action type
+ * @param {String}   type                      Triggered SCM event type ('pr' or 'repo')
+ * @param {String}   targetBranch              The branch against which commit is pushed
+ * @param {String}   pipelineBranch            The pipeline branch
+ * @param {String}   releaseName               SCM webhook release name
+ * @param {String}   tagName                   SCM webhook tag name
+ * @param {Boolean}  isReleaseOrTagFiltering   Check if the tag or release filtering or not
```
Comment (tkyi) [inline plugins/webhooks/index.js:46]:
```suggestion
 * @param {Boolean}  isReleaseOrTagFiltering   If the tag or release filtering is enabled
```

File: plugins/webhooks/index.js (hunk 4 above)
Comment (ibu1224) [reply plugins/webhooks/index.js]: Oh, I had a typo.  I fix it.

**Pull request merged**